jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+
+/**
+ * A Denial of Service Handler.
+ * <p>Protect from denial of service attacks by limiting the request rate from remote hosts</p>
+ */
+@ManagedObject("DoS Prevention Handler")
+public class DosHandler extends ConditionalHandler.ElseNext
+{
+    /**
+     * An id {@link Function} to create an ID from the remote address and port of a {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_ADDRESS_PORT = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote address of a {@link Request}
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_ADDRESS = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return inetSocketAddress.getAddress().toString();
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from the remote port of a {@link Request}.
+     * This can be useful if there is an untrusted intermediary, where the remote port can be a surrogate for the connection.
+     */
+    public static final Function<Request, String> ID_FROM_REMOTE_PORT = request ->
+    {
+        SocketAddress remoteSocketAddress = request.getConnectionMetaData().getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return Integer.toString(inetSocketAddress.getPort());
+        return remoteSocketAddress.toString();
+    };
+
+    /**
+     * An id {@link Function} to create an ID from {@link ConnectionMetaData#getId()} of the {@link Request}
+     */
+    public static final Function<Request, String> ID_CONNECTION = request -> request.getConnectionMetaData().getId();
+
+    /**
+     * An interface implemented to track and control the rate of requests for a specific ID.
+     */
+    public interface RateControl
+    {
+        /**
+         * Calculate if the rate is exceeded at the given time
+         * @param now The {@link NanoTime#now()} at which to calculate the rate
+         * @param addSample If {@code true} then a sample is added to the rate being calculated
+         * @return {@code true} if the rate is currently exceeded
+         */
+        boolean isRateExceeded(long now, boolean addSample);
+
+        /**
+         * Check if the tracker is now idle
+         * @param now The {@link NanoTime#now()} at which to calculate the rate
+         * @return {@code true} if the rate is currently near zero
+         */
+        boolean isIdle(long now);
```

Comment:
I do not understand this. Why do we need this method?
The javadocs refers to `Tracker`, not to `RateControl`.
`RateControl` should just be about the algorithm, not about idleness of an implementation detail?

Getting rid of this method would also get rid of the `addSample` parameter in `isRateExceeded()`, which seems a good thing.

Reply:
There are two ways that we call a Tracker: 1) a sample has occurred and needs to be added to the total and we need to check if the rate is exceeded; 2) time has passed and we need to check if the tracker is idle and thus can be removed.      So we cannot just have a single method.  API would be simpler with a `getRate` method, except we moved the `maxRate` into the tracker itself.     Also having a `addSample` and then `isRateExceeded` and `isIdle` methods would be a little cleaner, but some calculations would be needlessly repeated in the typical `addSample(); isRateExceeded();` calling pattern.

However, I think I can simplify the API a bit and move the avoidance of code duplication to the implementation... standby....

Reply:
I've simplified the contract.... not sure about the names...

Reply:
I don't understand why we need 2 methods. Idleness is determined by time, and the scheduler/cyclicTimeout takes care of that. One time expires, there is no need to ask again "are you idle" because of course it is.

About naming, I would move away from a getter-style `isRateExceeded()` and go for an event-like name such as `RateControl.onRequest()` (or `recordRequest()` or similar). In this way, it is clear that we are notifying the `RateControl` of an event that it should account for, removing "sample" from the name.

Reply:
@sbordet consider we have a Tracker that allows 100 requests per second and that in the first second 101 requests arrive and then no other requests arrive ever.   After that first second, `isRateExceeded` is true, because the rate is 101r/s.     Then (depending somewhat on how alpha is set) after a second, the rate will be 50r/s and the tracker is neither expired nor idle,   then after another second the rate is 33r/s, then 25/s, then 20/s.     Eventually the measured rate will reduce below some threshhold and the tracker will be seen as idle and deleted.  But depending on how alpha is set, that could be after 1 second or after 100s of seconds.

+1 for `onRequest` name.

Reply:
@gregw then is that a good rate control algorithm? That keeps around the Tracker for possibly long times?
For what characteristics was it chosen?

We don't need much beyond telling when the rate is exceeded, and telling when we're idle.

Calculating a smoothed average seems not that useful?